File: frontend/src/objects/player/preparation/player-preparation.ts
import { z } from "zod";

export const BirdStatsSchema = z.object({
  attack: z.number(),
  impact: z.number(),
  speed: z.number(),
});

export const BirdUpgradeSchema = z.object({
  birdType: z.string(),
  name: z.string(),
  summary: z.string(),
  previewImageUrl: z.string(),
  level: z.number(),
  maxLevel: z.number(),
  tier: z.number(),
  maxTier: z.number(),
  stats: BirdStatsSchema,
  skillName: z.string(),
  skillDescription: z.string(),
  nextTierSkillPreview: z.string().nullable(),
  nextCostCoins: z.number(),
  nextCostFragments: z.number(),
  source: z.enum(["system", "designer"]),
  authorId: z.string().nullable(),
  skills: z.unknown().nullable().optional(),
  modelImageUrl: z.string().nullable().optional(),
});

export const SlingshotUpgradeSchema = z.object({
  level: z.number(),
  maxLevel: z.number(),
  nextCostCoins: z.number(),
});

export const PreparationStateSchema = z.object({
  birds: z.array(BirdUpgradeSchema),
  slingshot: SlingshotUpgradeSchema,
  walletCoins: z.number(),
  walletFragments: z.number(),
});

export type BirdStats = z.infer<typeof BirdStatsSchema>;
export type BirdUpgradeState = z.infer<typeof BirdUpgradeSchema>;
export type SlingshotUpgradeState = z.infer<typeof SlingshotUpgradeSchema>;
export type PlayerPreparationState = z.infer<typeof PreparationStateSchema>;

File: backend/microservice/src/player/objects/preparation/PlayerPreparation.scala
package microservice.player.objects.preparation

import io.circe.Json

/** 玩家备战 API 领域对象。
  *
  * 定义：BirdUpgradeView、SlingshotUpgradeView、PlayerPreparationResponse。
  * 问题：准备页展示需要结构化升级/升阶信息。
  * 作用：由 PlayerPreparationSupport 组装。
  * 关联：frontend preparation schema；preparation 包下 APIMessage。
  */
final case class BirdStatsView(
  attack: Int,
  impact: Int,
  speed: Int
)

/** 单只鸟的备战升级视图：等级、阶位、属性、技能与下次升级花费。 */
final case class BirdUpgradeView(
  birdType: String,
  name: String,
  summary: String,
  previewImageUrl: String,
  level: Int,
  maxLevel: Int,
  tier: Int,
  maxTier: Int,
  stats: BirdStatsView,
  skillName: String,
  skillDescription: String,
  nextTierSkillPreview: Option[String],
  nextCostCoins: Int,
  nextCostFragments: Int,
  source: String,
  authorId: Option[String],
  skills: Option[Json] = None,
  modelImageUrl: Option[String] = None
)

/** 弹弓升级视图：当前等级与下次升级金币花费。 */
final case class SlingshotUpgradeView(
  level: Int,
  maxLevel: Int,
  nextCostCoins: Int
)

/** GET /player/preparation 的完整响应：全部鸟、弹弓与钱包余额摘要。 */
final case class PlayerPreparationResponse(
  birds: List[BirdUpgradeView],
  slingshot: SlingshotUpgradeView,
  walletCoins: Int,
  walletFragments: Int
)
